### PR TITLE
added mdm cluster resource

### PR DESCRIPTION
**Pull request by shenda1**
# Description
PR is raised to add new MDM resource

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 80% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Acceptance Tests

image

### CODE REVIEW

File: powerflex/provider/mdm_cluster_resource.go [hunk 1]
```diff
@@ -0,0 +1,884 @@
...
+
+	planSbMdmList := []models.StandByMdm{}
+	stateSbMdmList := []models.StandByMdm{}
+
+	// Populate planMdmList with standby MDMs defined in plan
+	diags := plan.StandByMdm.ElementsAs(ctx, &planSbMdmList, true)
+	dia.Append(diags...)
+	for _, mdm := range planSbMdmList {
+		if !mdm.ID.IsUnknown() && !mdm.Name.IsUnknown() {
+			planMdm[mdm.ID.ValueString()] = mdm.Name.ValueString()
```
Comment (rounak-adhikary) [inline powerflex/provider/mdm_cluster_resource.go:701]: Can we put a continue statement after this line and deindent the else block?

File: powerflex/provider/mdm_cluster_resource.go [hunk 2]
```diff
@@ -0,0 +1,884 @@
...
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error getting MDM cluster details: ",
+			"Error getting MDM cluster details: "+err.Error(),
+		)
+		return
+	}
+
+	// state, diag := helper.UpdateMdmClusterState(ctx, mdm_details, &plan, d.system.System.PerformanceProfile)
+	// diags.Append(diag...)
```
Comment (AnikaAgiwal2711) [inline powerflex/provider/mdm_cluster_resource.go:426]: Remove the commented code

File: powerflex/provider/mdm_cluster_resource.go [hunk 3]
```diff
@@ -0,0 +1,884 @@
...
+func (d *mdmClusterResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_mdm_cluster"
+}
+
+func (d *mdmClusterResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = MdmClusterResourceSchema
+}
+
+// MdmClusterResourceSchema defines the schema for Mdm resource
+var MdmClusterResourceSchema schema.Schema = schema.Schema{
```
Comment (AnikaAgiwal2711) [inline powerflex/provider/mdm_cluster_resource.go:69]: Please add the info in the description for the attributes which cannot be updated

File: powerflex/helper/mdm_cluster_helper.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+			ID:            types.StringValue(mdm.ID),
+			Name:          types.StringValue(mdm.Name),
+			Port:          types.Int64Value(int64(mdm.Port)),
+			Ips:           mdmIps,
+			ManagementIps: mgmtIps,
+		}
+	}
+
+	tbMdmDetails := make([]models.Mdm, 0)
+	// for _, mdm := range plan.TieBreakerMdm {
```
Comment (AnikaAgiwal2711) [inline powerflex/helper/mdm_cluster_helper.go:206]: same here

File: powerflex/provider/mdm_cluster_resource.go [hunk 5]
```diff
@@ -0,0 +1,884 @@
...
+	system, err := helper.GetFirstSystem(d.client)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to Read Powerflex System",
+			err.Error(),
+		)
+		return
+	}
+
+	state, dgs = helper.UpdateMdmClusterState(ctx, mdmDetails, &plan, system.System.PerformanceProfile)
```
[the remark this answers is not in the document]
Comment (shenda1) [reply powerflex/provider/mdm_cluster_resource.go:530]: As discussed, this is a new system type.

File: powerflex/provider/mdm_cluster_resource.go [hunk 6]
```diff
@@ -0,0 +1,884 @@
...
+		return
+	}
+
+	// state, diag := helper.UpdateMdmClusterState(ctx, mdm_details, &plan, d.system.System.PerformanceProfile)
+	// diags.Append(diag...)
+
+	// Get the IP with IDs as value
+	ipmap := helper.GetMdmIPMap(mdmDetails)
+
+	// Populate Primary MDM ID if IP is provided
```
[the remark this answers is not in the document]
Comment (shenda1) [reply powerflex/provider/mdm_cluster_resource.go:431]: As discussed, method is called from update and create.

File: powerflex/provider/mdm_cluster_resource.go (hunk 1 above)
Comment (shenda1) [reply powerflex/provider/mdm_cluster_resource.go:701]: As discussed, continue statement not needed anymore.

File: powerflex/provider/mdm_cluster_resource.go (hunk 2 above)
Comment (shenda1) [reply powerflex/provider/mdm_cluster_resource.go:426]: done

File: powerflex/helper/mdm_cluster_helper.go (hunk 4 above)
Comment (shenda1) [reply powerflex/helper/mdm_cluster_helper.go:206]: done

File: powerflex/helper/mdm_cluster_helper.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+			ID:            types.StringValue(mdm.ID),
+			Name:          types.StringValue(mdm.Name),
+			Port:          types.Int64Value(int64(mdm.Port)),
+			Ips:           mdmIps,
+			ManagementIps: mgmtIps,
+		}
+	}
+
+	secondaryMdmDetails := make([]models.Mdm, 0)
+	// for _, mdm := range plan.SecondaryMdm {
```
[the remark this answers is not in the document]
Comment (shenda1) [reply powerflex/helper/mdm_cluster_helper.go:168]: done

**Review by rounak-adhikary — APPROVED** [review]
LGTM

**Review by Krunal-Thakkar — DISMISSED** [review]

**Comment by rounak-adhikary** [comment]
run e2e

**Comment by Krunal-Thakkar** [comment]
run e2e test

**Comment by shenda1** [comment]
run e2e test

**Comment by shenda1** [comment]
run e2e test

**Comment by shenda1** [comment]
run e2e test

**Comment by shenda1** [comment]
run e2e test

**Comment by shenda1** [comment]
run e2e test

**Review by rounak-adhikary — APPROVED** [review]
LGTM

**Review by Krunal-Thakkar — APPROVED** [review]

**Pull request merged**